discovery/xds/client.go

```diff
@@ -0,0 +1,229 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/version"
+)
+
+var userAgent = fmt.Sprintf("Prometheus/%s", version.Version)
+
+// ResourceClient exposes the xDS protocol for a single resource type.
+// See https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#rest-json-polling-subscriptions .
+type ResourceClient interface {
+	// ResourceTypeURL is the type URL of the resource.
+	ResourceTypeURL() string
+
+	// Server is the xDS Management server.
+	Server() string
+
+	// Fetch requests the latest view of the entire resource state.
+	// If no updates have been made since the last request, the response will be nil.
+	Fetch(ctx context.Context) (*v3.DiscoveryResponse, error)
+
+	// ID returns the ID of the client that is sent to the xDS server.
+	ID() string
+
+	// Close releases currently held resources.
+	Close()
+}
+
+type HTTPResourceClient struct {
+	client *http.Client
+	config *HTTPResourceClientConfig
+
+	// endpoint is the fully-constructed xDS HTTP endpoint.
+	endpoint string
+	// Caching.
+	latestVersion string
+	latestNonce   string
+}
+
+type HTTPResourceClientConfig struct {
+	// HTTP config.
+	config.HTTPClientConfig
+
+	Name string
+
+	// ExtraQueryParams are extra query parameters to attach to the request URL.
+	ExtraQueryParams url.Values
+
+	// General xDS config.
+
+	// The timeout for a single fetch request.
+	Timeout time.Duration
+
+	// Type is the xds type, e.g., clusters
+	// which is used in the discovery POST request.
+	ResourceType string
+	// ResourceTypeURL is the Google type url for the resource, e.g., type.googleapis.com/envoy.api.v2.Cluster.
+	ResourceTypeURL string
+	// Server is the xDS management server.
+	Server string
+	// ClientID is used to identify the client with the management server.
+	ClientID string
+}
+
+func NewHTTPResourceClient(conf *HTTPResourceClientConfig, protocolVersion ProtocolVersion) (*HTTPResourceClient, error) {
+	if protocolVersion != ProtocolV3 {
+		return nil, errors.New("only the v3 protocol is supported")
+	}
+
+	if len(conf.Server) == 0 {
+		return nil, errors.New("empty xDS server")
+	}
+
+	serverURL, err := url.Parse(conf.Server)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointURL, err := makeXDSResourceHTTPEndpointURL(protocolVersion, serverURL, conf.ResourceType)
+	if err != nil {
+		return nil, err
+	}
+
+	if conf.ExtraQueryParams != nil {
+		endpointURL.RawQuery = conf.ExtraQueryParams.Encode()
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, conf.Name, config.WithHTTP2Disabled(), config.WithIdleConnTimeout(conf.Timeout))
```

Comment:
We could directly call NewClientFromConfig

```suggestion
	rt, err := config.NewClientFromConfig(conf.HTTPClientConfig, conf.Name, config.WithHTTP2Disabled(), config.WithIdleConnTimeout(conf.Timeout))
```